{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\\[_In case you’re unable to see the atoti visualizations in GitHub, try viewing the notebook in [nbviewer](https://nbviewer.org/github/atoti/notebooks/blob/master/notebooks/airline-industry/main.ipynb)._]\n",
    "\n",
    "## OLAP operations with atoti for use case from Airline Industry\n",
    "From the vast domain of the aviation sector, we present a use case pertaining to performance of the airline industry during the covid-19 pandemic. The notebook illustrates the smooth blending of olap operations with atoti to simulate/model multiple *“what if”* scenarios easily, to gain crucial insights."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dataset used:\n",
    "The simulated dataset has the three dimensions as :-\n",
    "* airlines - Contains some important information pertaining to different  airlines, types of aircrafts, categories of booking class, seats availability w.r.t booking class, daily departures  among others.\n",
    "* routes - The routes operated by each airline and distance between the source and destination in kilometers for each airline etc.\n",
    "* date - The occurrence of real world events such as booking a ticket can be measured at different levels of hierarchy such as year, month, date, etc.\n",
    "\n",
    "In addition to that we have the fact table(which has the measures) \n",
    "* salesdata -  Here all details containing the tickets sold by a particular airline along specific routes are filled which can be used to get insights about some business performance metrics such as revenue generated by a specific airline, revenue generated along specific routes, revenue generated with respect to a particular booking category, etc.\n",
    "\n",
    "* The use case considers seven airlines and 24 routes for each of the seven airlines."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import python libraries\n",
    "import datetime\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "# Import atoti library\n",
    "import atoti as tt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part I - Loading, joining the data and creating cubes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load sales data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.6.5'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tt.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Expected one of ('true', '1') or ('false', '0') but received atotijavaexception.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mPy4JJavaError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti\\_java_api.py:128\u001b[0m, in \u001b[0;36m_enhance_py4j_errors.<locals>.wrapped_method\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m    127\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 128\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m function(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    129\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m Py4JJavaError \u001b[38;5;28;01mas\u001b[39;00m java_exception:\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti\\_java_api.py:313\u001b[0m, in \u001b[0;36mJavaApi.start_application\u001b[1;34m(self, config)\u001b[0m\n\u001b[0;32m    312\u001b[0m json_config \u001b[38;5;241m=\u001b[39m serialize_config_to_json(config)\n\u001b[1;32m--> 313\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjava_session\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstartServer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mjson_config\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\py4j\\java_gateway.py:1321\u001b[0m, in \u001b[0;36mJavaMember.__call__\u001b[1;34m(self, *args)\u001b[0m\n\u001b[0;32m   1320\u001b[0m answer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgateway_client\u001b[38;5;241m.\u001b[39msend_command(command)\n\u001b[1;32m-> 1321\u001b[0m return_value \u001b[38;5;241m=\u001b[39m \u001b[43mget_return_value\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m   1322\u001b[0m \u001b[43m    \u001b[49m\u001b[43manswer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgateway_client\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtarget_id\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1324\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m temp_arg \u001b[38;5;129;01min\u001b[39;00m temp_args:\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\py4j\\protocol.py:326\u001b[0m, in \u001b[0;36mget_return_value\u001b[1;34m(answer, gateway_client, target_id, name)\u001b[0m\n\u001b[0;32m    325\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m answer[\u001b[38;5;241m1\u001b[39m] \u001b[38;5;241m==\u001b[39m REFERENCE_TYPE:\n\u001b[1;32m--> 326\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m Py4JJavaError(\n\u001b[0;32m    327\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAn error occurred while calling \u001b[39m\u001b[38;5;132;01m{0}\u001b[39;00m\u001b[38;5;132;01m{1}\u001b[39;00m\u001b[38;5;132;01m{2}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39m\n\u001b[0;32m    328\u001b[0m         \u001b[38;5;28mformat\u001b[39m(target_id, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m, name), value)\n\u001b[0;32m    329\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[1;31mPy4JJavaError\u001b[0m: An error occurred while calling t.startServer.\n: org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Tomcat\r\n\tat org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:163)\r\n\tat org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:577)\r\n\tat org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145)\r\n\tat org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730)\r\n\tat org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:412)\r\n\tat org.springframework.boot.SpringApplication.run(SpringApplication.java:302)\r\n\tat org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:164)\r\n\tat io.atoti.util.impl.ServerUtils.startApplication(ServerUtils.java:65)\r\n\tat io.atoti.util.impl.ServerManager.startServer(ServerManager.java:64)\r\n\tat io.atoti.SessionImpl.startServer(SessionImpl.java:208)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\r\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\r\n\tat py4j.Gateway.invoke(Gateway.java:282)\r\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\r\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\r\n\tat py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)\r\n\tat py4j.ClientServerConnection.run(ClientServerConnection.java:106)\r\n\tat java.base/java.lang.Thread.run(Unknown Source)\r\nCaused by: org.springframework.boot.web.server.WebServerException: Unable to start embedded Tomcat\r\n\tat org.springframework.boot.web.embedded.tomcat.TomcatWebServer.initialize(TomcatWebServer.java:142)\r\n\tat org.springframework.boot.web.embedded.tomcat.TomcatWebServer.<init>(TomcatWebServer.java:104)\r\n\tat org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getTomcatWebServer(TomcatServletWebServerFactory.java:473)\r\n\tat org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory.getWebServer(TomcatServletWebServerFactory.java:206)\r\n\tat org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.createWebServer(ServletWebServerApplicationContext.java:182)\r\n\tat org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.onRefresh(ServletWebServerApplicationContext.java:160)\r\n\t... 21 more\r\nCaused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValueFilter' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.servlet.impl.ContextValueFilter]: Factory method 'contextValueFilter' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValuePropagator' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValuePropagator]: Factory method 'contextValuePropagator' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValueManager' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValueManager]: Factory method 'contextValueManager' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'activePivotContentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:486)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)\r\n\tat org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:213)\r\n\tat org.springframework.boot.web.servlet.ServletContextInitializerBeans.getOrderedBeansOfType(ServletContextInitializerBeans.java:212)\r\n\tat org.springframework.boot.web.servlet.ServletContextInitializerBeans.addAsRegistrationBean(ServletContextInitializerBeans.java:175)\r\n\tat org.springframework.boot.web.servlet.ServletContextInitializerBeans.addAsRegistrationBean(ServletContextInitializerBeans.java:170)\r\n\tat org.springframework.boot.web.servlet.ServletContextInitializerBeans.addAdaptableBeans(ServletContextInitializerBeans.java:155)\r\n\tat org.springframework.boot.web.servlet.ServletContextInitializerBeans.<init>(ServletContextInitializerBeans.java:87)\r\n\tat org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.getServletContextInitializerBeans(ServletWebServerApplicationContext.java:260)\r\n\tat org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.selfInitialize(ServletWebServerApplicationContext.java:234)\r\n\tat org.springframework.boot.web.embedded.tomcat.TomcatStarter.onStartup(TomcatStarter.java:53)\r\n\tat org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5219)\r\n\tat org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)\r\n\tat org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1396)\r\n\tat org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1386)\r\n\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\r\n\tat org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)\r\n\tat java.base/java.util.concurrent.AbstractExecutorService.submit(Unknown Source)\r\n\tat org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:919)\r\n\tat org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:835)\r\n\tat org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)\r\n\tat org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1396)\r\n\tat org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1386)\r\n\tat java.base/java.util.concurrent.FutureTask.run(Unknown Source)\r\n\tat org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)\r\n\tat java.base/java.util.concurrent.AbstractExecutorService.submit(Unknown Source)\r\n\tat org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:919)\r\n\tat org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:263)\r\n\tat org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)\r\n\tat org.apache.catalina.core.StandardService.startInternal(StandardService.java:432)\r\n\tat org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)\r\n\tat org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:927)\r\n\tat org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)\r\n\tat org.apache.catalina.startup.Tomcat.start(Tomcat.java:486)\r\n\tat org.springframework.boot.web.embedded.tomcat.TomcatWebServer.initialize(TomcatWebServer.java:123)\r\n\t... 26 more\r\nCaused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.servlet.impl.ContextValueFilter]: Factory method 'contextValueFilter' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValuePropagator' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValuePropagator]: Factory method 'contextValuePropagator' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValueManager' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValueManager]: Factory method 'contextValueManager' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'activePivotContentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)\r\n\t... 67 more\r\nCaused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValuePropagator' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValuePropagator]: Factory method 'contextValuePropagator' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValueManager' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValueManager]: Factory method 'contextValueManager' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'activePivotContentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:486)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)\r\n\tat org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.resolveBeanReference(ConfigurationClassEnhancer.java:362)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:334)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.contextValuePropagator(<generated>)\r\n\tat io.atoti.spring.ActivePivotConfig.contextValueFilter(ActivePivotConfig.java:78)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.CGLIB$contextValueFilter$4(<generated>)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4$$FastClassBySpringCGLIB$$d3f6ae8b.invoke(<generated>)\r\n\tat org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.contextValueFilter(<generated>)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)\r\n\t... 68 more\r\nCaused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValuePropagator]: Factory method 'contextValuePropagator' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValueManager' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValueManager]: Factory method 'contextValueManager' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'activePivotContentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)\r\n\t... 91 more\r\nCaused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contextValueManager' defined in io.atoti.spring.ActivePivotConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValueManager]: Factory method 'contextValueManager' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'activePivotContentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:486)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)\r\n\tat org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.resolveBeanReference(ConfigurationClassEnhancer.java:362)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:334)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.contextValueManager(<generated>)\r\n\tat io.atoti.spring.ActivePivotConfig.contextValuePropagator(ActivePivotConfig.java:90)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.CGLIB$contextValuePropagator$0(<generated>)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4$$FastClassBySpringCGLIB$$d3f6ae8b.invoke(<generated>)\r\n\tat org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.contextValuePropagator(<generated>)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)\r\n\t... 92 more\r\nCaused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.quartetfs.biz.pivot.security.impl.ContextValueManager]: Factory method 'contextValueManager' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'activePivotContentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)\r\n\t... 115 more\r\nCaused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'activePivotContentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:486)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)\r\n\tat org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.resolveBeanReference(ConfigurationClassEnhancer.java:362)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:334)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.activePivotContentService(<generated>)\r\n\tat io.atoti.spring.ActivePivotConfig.contextValueManager(ActivePivotConfig.java:61)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.CGLIB$contextValueManager$2(<generated>)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4$$FastClassBySpringCGLIB$$d3f6ae8b.invoke(<generated>)\r\n\tat org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)\r\n\tat io.atoti.spring.ActivePivotConfig$$EnhancerBySpringCGLIB$$dd7c1fa4.contextValueManager(<generated>)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)\r\n\t... 116 more\r\nCaused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.pivot.content.IActivePivotContentService]: Factory method 'activePivotContentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)\r\n\t... 139 more\r\nCaused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'contentService' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:638)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)\r\n\tat org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.resolveBeanReference(ConfigurationClassEnhancer.java:362)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:334)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.contentService(<generated>)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig.activePivotContentService(JdbcContentServiceConfig.java:63)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.CGLIB$activePivotContentService$2(<generated>)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557$$FastClassBySpringCGLIB$$58af597a.invoke(<generated>)\r\n\tat org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.activePivotContentService(<generated>)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)\r\n\t... 140 more\r\nCaused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.qfs.content.service.IContentService]: Factory method 'contentService' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)\r\n\t... 163 more\r\nCaused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'jdbcDataSource' defined in io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:486)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)\r\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)\r\n\tat org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)\r\n\tat org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.resolveBeanReference(ConfigurationClassEnhancer.java:362)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:334)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.jdbcDataSource(<generated>)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig.createHibernateContentServer(JdbcContentServiceConfig.java:87)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig.getContentServiceRequiringUiInitialization(JdbcContentServiceConfig.java:49)\r\n\tat io.atoti.patachou.contentserver.cfg.AContentServiceConfig.contentService(AContentServiceConfig.java:52)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.CGLIB$contentService$3(<generated>)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557$$FastClassBySpringCGLIB$$58af597a.invoke(<generated>)\r\n\tat org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.contentService(<generated>)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)\r\n\t... 164 more\r\nCaused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [javax.sql.DataSource]: Factory method 'jdbcDataSource' threw exception; nested exception is com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)\r\n\tat org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)\r\n\t... 189 more\r\nCaused by: com.zaxxer.hikari.pool.HikariPool$PoolInitializationException: Failed to initialize pool: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat com.zaxxer.hikari.pool.HikariPool.throwPoolInitializationException(HikariPool.java:596)\r\n\tat com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:582)\r\n\tat com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)\r\n\tat com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:81)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig.jdbcDataSource(JdbcContentServiceConfig.java:103)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.CGLIB$jdbcDataSource$1(<generated>)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557$$FastClassBySpringCGLIB$$58af597a.invoke(<generated>)\r\n\tat org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)\r\n\tat org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)\r\n\tat io.atoti.patachou.contentserver.cfg.JdbcContentServiceConfig$$EnhancerBySpringCGLIB$$a909b557.jdbcDataSource(<generated>)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\r\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)\r\n\tat java.base/java.lang.reflect.Method.invoke(Unknown Source)\r\n\tat org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)\r\n\t... 190 more\r\nCaused by: org.h2.jdbc.JdbcSQLNonTransientException: General error: \"The write format 1 is smaller than the supported format 2 [2.1.210/5]\" [50000-210]\r\n\tat org.h2.message.DbException.getJdbcSQLException(DbException.java:573)\r\n\tat org.h2.message.DbException.getJdbcSQLException(DbException.java:496)\r\n\tat org.h2.message.DbException.get(DbException.java:216)\r\n\tat org.h2.mvstore.db.Store.convertMVStoreException(Store.java:166)\r\n\tat org.h2.mvstore.db.Store.<init>(Store.java:140)\r\n\tat org.h2.engine.Database.<init>(Database.java:324)\r\n\tat org.h2.engine.Engine.openSession(Engine.java:92)\r\n\tat org.h2.engine.Engine.openSession(Engine.java:222)\r\n\tat org.h2.engine.Engine.createSession(Engine.java:201)\r\n\tat org.h2.engine.SessionRemote.connectEmbeddedOrServer(SessionRemote.java:338)\r\n\tat org.h2.jdbc.JdbcConnection.<init>(JdbcConnection.java:122)\r\n\tat org.h2.Driver.connect(Driver.java:59)\r\n\tat com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:121)\r\n\tat com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:364)\r\n\tat com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)\r\n\tat com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:476)\r\n\tat com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)\r\n\t... 203 more\r\nCaused by: org.h2.mvstore.MVStoreException: The write format 1 is smaller than the supported format 2 [2.1.210/5]\r\n\tat org.h2.mvstore.DataUtils.newMVStoreException(DataUtils.java:1004)\r\n\tat org.h2.mvstore.MVStore.getUnsupportedWriteFormatException(MVStore.java:1059)\r\n\tat org.h2.mvstore.MVStore.readStoreHeader(MVStore.java:878)\r\n\tat org.h2.mvstore.MVStore.<init>(MVStore.java:455)\r\n\tat org.h2.mvstore.MVStore$Builder.open(MVStore.java:4056)\r\n\tat org.h2.mvstore.db.Store.<init>(Store.java:129)\r\n\t... 215 more\r\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[1;32mIn [2]\u001b[0m, in \u001b[0;36m<cell line: 2>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# creating a session\u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m session \u001b[38;5;241m=\u001b[39m \u001b[43mtt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mSession\u001b[49m\u001b[43m(\u001b[49m\u001b[43muser_content_storage\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m./content\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti\\_runtime_type_checking_utils.py:179\u001b[0m, in \u001b[0;36m_TypecheckWrapperFactory.create_wrapper.<locals>.typechecked_func_wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    176\u001b[0m typeguard\u001b[38;5;241m.\u001b[39mcheck_argument_types(memo)\n\u001b[0;32m    178\u001b[0m \u001b[38;5;66;03m# Call the actual function.\u001b[39;00m\n\u001b[1;32m--> 179\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti_core\\doc.py:21\u001b[0m, in \u001b[0;36mdoc.<locals>.decorator.<locals>.wrapper\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m     19\u001b[0m \u001b[38;5;129m@wraps\u001b[39m(function)\n\u001b[0;32m     20\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrapper\u001b[39m(\u001b[38;5;241m*\u001b[39margs: _P\u001b[38;5;241m.\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: _P\u001b[38;5;241m.\u001b[39mkwargs) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Any:\n\u001b[1;32m---> 21\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m function(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti\\session.py:231\u001b[0m, in \u001b[0;36mSession.__init__\u001b[1;34m(self, name, app_extensions, authentication, branding, certificate_authority, client_certificate, extra_jars, https, i18n, java_options, jwt, logging, port, same_site, user_content_storage, **kwargs)\u001b[0m\n\u001b[0;32m    228\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    229\u001b[0m     _sessions\u001b[38;5;241m.\u001b[39m_clear_duplicate_sessions(name)\n\u001b[1;32m--> 231\u001b[0m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__init__\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[0;32m    232\u001b[0m \u001b[43m    \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    233\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconfig\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconfig\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    234\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdistributed\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[0;32m    235\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcertificate_authority\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcertificate_authority\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    236\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    238\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_warn_if_license_about_to_expire()\n\u001b[0;32m    239\u001b[0m \u001b[38;5;66;03m# https://github.com/python/mypy/issues/12183\u001b[39;00m\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti\\_local_session.py:83\u001b[0m, in \u001b[0;36mLocalSession.__init__\u001b[1;34m(self, config, distributed, certificate_authority, name)\u001b[0m\n\u001b[0;32m     80\u001b[0m     plugin\u001b[38;5;241m.\u001b[39minit_session(\u001b[38;5;28mself\u001b[39m)\n\u001b[0;32m     82\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m---> 83\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_start_application\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     84\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m AtotiJavaException \u001b[38;5;28;01mas\u001b[39;00m ave:\n\u001b[0;32m     85\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m AtotiException(\n\u001b[0;32m     86\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mave\u001b[38;5;241m.\u001b[39mjava_traceback\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     87\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAn error occurred while configuring the session.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     88\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe logs are available at \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlogs_path\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     89\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti\\_local_session.py:155\u001b[0m, in \u001b[0;36mLocalSession._start_application\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    153\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_start_application\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    154\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 155\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_java_api\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mstart_application\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_config\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    156\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m AtotiJavaException \u001b[38;5;28;01mas\u001b[39;00m error:\n\u001b[0;32m    157\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[0;32m    158\u001b[0m             \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_config\u001b[38;5;241m.\u001b[39muser_content_storage, UserContentStorageConfig)\n\u001b[0;32m    159\u001b[0m             \u001b[38;5;129;01mand\u001b[39;00m is_h2_v1_error(\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    163\u001b[0m             \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_server_subprocess\n\u001b[0;32m    164\u001b[0m         ):\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti\\_java_api.py:132\u001b[0m, in \u001b[0;36m_enhance_py4j_errors.<locals>.wrapped_method\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m    128\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m function(\u001b[38;5;28mself\u001b[39m, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    129\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m Py4JJavaError \u001b[38;5;28;01mas\u001b[39;00m java_exception:\n\u001b[0;32m    130\u001b[0m     cause \u001b[38;5;241m=\u001b[39m (\n\u001b[0;32m    131\u001b[0m         \u001b[38;5;28mstr\u001b[39m(java_exception)\n\u001b[1;32m--> 132\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mget_env_flag\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_ATOTI_VERBOSE_JAVA_EXCEPTIONS\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    133\u001b[0m         \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_throwable_root_cause(java_exception\u001b[38;5;241m.\u001b[39mjava_exception)\n\u001b[0;32m    134\u001b[0m     )\n\u001b[0;32m    135\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m AtotiJavaException(\n\u001b[0;32m    136\u001b[0m         cause,\n\u001b[0;32m    137\u001b[0m         java_traceback\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mstr\u001b[39m(java_exception),\n\u001b[0;32m    138\u001b[0m         java_exception\u001b[38;5;241m=\u001b[39mjava_exception,\n\u001b[0;32m    139\u001b[0m     ) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28mNone\u001b[39m\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti_core\\get_env_flag.py:7\u001b[0m, in \u001b[0;36mget_env_flag\u001b[1;34m(variable_name)\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_env_flag\u001b[39m(variable_name: \u001b[38;5;28mstr\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n\u001b[1;32m----> 7\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mstr_to_bool\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m      8\u001b[0m \u001b[43m        \u001b[49m\u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menviron\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m      9\u001b[0m \u001b[43m            \u001b[49m\u001b[43mvariable_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m     10\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;66;43;03m# The default is not configurable because it's simpler if `absence of the flag <=> the flag is false`.\u001b[39;49;00m\n\u001b[0;32m     11\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;28;43mstr\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m     12\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     13\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mC:\\Apps\\miniconda3\\envs\\atoti\\lib\\site-packages\\atoti_core\\str_to_bool.py:16\u001b[0m, in \u001b[0;36mstr_to_bool\u001b[1;34m(value)\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m value \u001b[38;5;129;01min\u001b[39;00m _FALSE_VALUES:\n\u001b[0;32m     14\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m---> 16\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m     17\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mExpected one of \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m_TRUE_VALUES\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m or \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m_FALSE_VALUES\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m but received \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mvalue\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m     18\u001b[0m )\n",
      "\u001b[1;31mValueError\u001b[0m: Expected one of ('true', '1') or ('false', '0') but received atotijavaexception."
     ]
    }
   ],
   "source": [
    "# creating a session\n",
    "session = tt.Session(user_content_storage= \"./content\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session.link()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading salesdata\n",
    "salesdata = pd.read_csv(\"data/salesdata.csv\")\n",
    "salesdata_table = session.read_pandas(\n",
    "    salesdata,\n",
    "    keys=[\"SalesId\", \"Date\", \"RouteId\"],\n",
    "    table_name=\"salesdata_table\",\n",
    "    types={\n",
    "        \"SalesId\": tt.type.STRING,\n",
    "        \"Date\": tt.type.STRING,\n",
    "        \"RouteId\": tt.type.STRING,\n",
    "    },\n",
    ")\n",
    "salesdata_table.head(8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading airlines dataset to skip the index column and then store it in atoti store\n",
    "\n",
    "airline_df = pd.read_csv(\"data/airlines.csv\", index_col=0)\n",
    "airlines_table = session.read_pandas(\n",
    "    airline_df, keys=[\"AirlineId\"], table_name=\"airlines_table\"\n",
    ")\n",
    "airline_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading routes dataset\n",
    "\n",
    "routes_table = session.read_csv(\"data/routes.csv\", keys=[\"RouteId\"])\n",
    "routes_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading dates dataset\n",
    "df = pd.read_csv(\"data/date.csv\")\n",
    "df = df.astype(str)\n",
    "dates_table = session.read_pandas(df, keys=[\"Date\"], table_name=\"dates_table\")\n",
    "dates_table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Join the data tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we join the newly created stores to the sales store\n",
    "\n",
    "salesdata_table.join(airlines_table, mapping={\"AirlineId\": \"AirlineId\"})\n",
    "\n",
    "salesdata_table.join(routes_table, mapping={\"RouteId\": \"RouteId\"})\n",
    "\n",
    "salesdata_table.join(dates_table, mapping={\"Date\": \"Date\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating the cube"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cube = session.create_cube(salesdata_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session.tables.schema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# defining measures, level and hierarchy\n",
    "m = cube.measures\n",
    "l = cube.levels\n",
    "h = cube.hierarchies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Defining New Custom Measure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we saw in the salesdata table we have multiple active airlines for a day, but we may need daily aggregate revenue.So we will define a new measure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m[\"Total Daily Revenue\"] = tt.agg.sum(\n",
    "    tt.value(salesdata_table[\"Total\"]),\n",
    "    scope=tt.scope.origin(l[\"Date\"]),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PART II: Visualizations\n",
    "\n",
    "We Present the following insights by implementing olap operations (slice, dice and drill down) with atoti."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing yearly and quarterly Revenue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first insight visualised through pie-chart clearly shows that covid-19 had a negative impact on the airline industry as we see fall in revenue for year 2021 as compared to 2020. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "atoti": {
     "widget": {
      "mapping": {
       "horizontalSubplots": [],
       "sliceBy": [
        "[dates_table].[Year].[Year]"
       ],
       "values": [
        "[Measures].[Total Daily Revenue]"
       ],
       "verticalSubplots": []
      },
      "query": {
       "mdx": "SELECT NON EMPTY {[Measures].[Total Daily Revenue]} ON COLUMNS, NON EMPTY Hierarchize(Descendants({[dates_table].[Year].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS FROM [salesdata_table] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
       "updateMode": "once"
      },
      "serverKey": "default",
      "widgetKey": "plotly-pie-chart"
     }
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "session.visualize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Furthermore if we see the performance of year 2020 across all four quarters we see that quarter three was the one which was negatively impacted as compared to other quarters."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "atoti": {
     "widget": {
      "filters": [
       "[dates_table].[Year].[AllMember].[2020]"
      ],
      "mapping": {
       "horizontalSubplots": [],
       "sliceBy": [
        "[dates_table].[Quarter].[Quarter]"
       ],
       "values": [
        "[Measures].[Total Daily Revenue]"
       ],
       "verticalSubplots": []
      },
      "query": {
       "mdx": "SELECT NON EMPTY {[Measures].[Total Daily Revenue]} ON COLUMNS, NON EMPTY Hierarchize(Descendants({[dates_table].[Quarter].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS FROM [salesdata_table] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
       "updateMode": "once"
      },
      "serverKey": "default",
      "widgetKey": "plotly-pie-chart"
     }
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "session.visualize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing quarterly revenue for four different airlines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This insight presents a comparative analysis of four different airlines that were active for the year 2020 and the revenue generated by each airline in these four quarters. We see that Allegiant Airlines business was negatively impacted in quarter 4. This can be used to simulate drill down operation as year → quarter with aggregating across routes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "atoti": {
     "height": 441,
     "widget": {
      "filters": [
       "[dates_table].[Year].[AllMember].[2020]"
      ],
      "mapping": {
       "horizontalSubplots": [],
       "stackBy": [
        "[dates_table].[Quarter].[Quarter]",
        "ALL_MEASURES"
       ],
       "values": [
        "[Measures].[Total Daily Revenue]"
       ],
       "verticalSubplots": [],
       "xAxis": [
        "[airlines_table].[AirlineName].[AirlineName]"
       ]
      },
      "query": {
       "mdx": "SELECT NON EMPTY Crossjoin(Hierarchize(Descendants({[dates_table].[Quarter].[AllMember]}, 1, SELF_AND_BEFORE)), {[Measures].[Total Daily Revenue]}) ON COLUMNS, NON EMPTY Hierarchize(Descendants({[airlines_table].[AirlineName].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS FROM [salesdata_table] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
       "updateMode": "once"
      },
      "serverKey": "default",
      "widgetKey": "plotly-stacked-column-chart"
     }
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "session.visualize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing quarterly revenue for four different airlines and drilling down across routes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we again have a comparative analysis of the above four airlines but now the revenue generated is shown along each specific route that the airlines were active in for all quarters.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "atoti": {
     "widget": {
      "filters": [
       "{[dates_table].[Quarter].[AllMember].[1], [dates_table].[Quarter].[AllMember].[2]}",
       "[dates_table].[Year].[AllMember].[2020]"
      ],
      "mapping": {
       "horizontalSubplots": [
        "[airlines_table].[AirlineName].[AirlineName]"
       ],
       "splitBy": [
        "[salesdata_table].[RouteId].[RouteId]",
        "ALL_MEASURES"
       ],
       "values": [
        "[Measures].[Total Daily Revenue]"
       ],
       "verticalSubplots": [],
       "xAxis": []
      },
      "query": {
       "mdx": "SELECT NON EMPTY Crossjoin(Hierarchize(Descendants({[salesdata_table].[RouteId].[AllMember]}, 1, SELF_AND_BEFORE)), {[Measures].[Total Daily Revenue]}, Hierarchize(Descendants({[airlines_table].[AirlineName].[AllMember]}, 1, SELF_AND_BEFORE))) ON COLUMNS FROM [salesdata_table] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
       "updateMode": "once"
      },
      "serverKey": "default",
      "widgetKey": "plotly-clustered-column-chart"
     }
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "session.visualize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing quarterly revenue for a single airlines across routes by slicing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We further slice the cube for “American Airlines” to show the total revenue across the four quarters along 12 distinct routes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "atoti": {
     "widget": {
      "filters": [
       "[airlines_table].[AirlineName].[AllMember].[American (American Airlines)]",
       "[dates_table].[Year].[AllMember].[2020]"
      ],
      "mapping": {
       "horizontalSubplots": [
        "[dates_table].[Quarter].[Quarter]"
       ],
       "stackBy": [
        "ALL_MEASURES"
       ],
       "values": [
        "[Measures].[Total Daily Revenue]"
       ],
       "verticalSubplots": [],
       "xAxis": [
        "[salesdata_table].[RouteId].[RouteId]"
       ]
      },
      "query": {
       "mdx": "SELECT NON EMPTY Crossjoin({[Measures].[Total Daily Revenue]}, Hierarchize(Descendants({[dates_table].[Quarter].[AllMember]}, 1, SELF_AND_BEFORE))) ON COLUMNS, NON EMPTY Hierarchize(Descendants({[salesdata_table].[RouteId].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS FROM [salesdata_table] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
       "updateMode": "once"
      },
      "serverKey": "default",
      "widgetKey": "plotly-stacked-column-chart"
     }
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "session.visualize(\"American Airlines\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing revenue for a month by booking classes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This insight shows the revenue generated by a particular booking class. We consider three categories such as business class, premium economy and economy.\n",
    "This is a dice operation as we dice it for year (2020), month (May) and Route(route1) for a specific airline(American Airline)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "salesdata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "airlines_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "atoti": {
     "height": 522,
     "widget": {
      "filters": [
       "[airlines_table].[AirlineName].[AllMember].[American (American Airlines)]",
       "[salesdata_table].[RouteId].[AllMember].[r002]",
       "[dates_table].[Month].[AllMember].[6]",
       "[dates_table].[Year].[AllMember].[2020]"
      ],
      "mapping": {
       "horizontalSubplots": [],
       "splitBy": [
        "ALL_MEASURES"
       ],
       "values": [
        "[Measures].[BookingsBusinessClass.SUM]",
        "[Measures].[BookingsPremiumEconomyClass.SUM]",
        "[Measures].[BookingsEconomyClass.SUM]"
       ],
       "verticalSubplots": [],
       "xAxis": [
        "[salesdata_table].[Date].[Date]"
       ]
      },
      "query": {
       "mdx": "SELECT NON EMPTY Hierarchize(Descendants({[salesdata_table].[Date].[AllMember]}, 1, SELF_AND_BEFORE)) ON ROWS, NON EMPTY {[Measures].[BookingsBusinessClass.SUM], [Measures].[BookingsPremiumEconomyClass.SUM], [Measures].[BookingsEconomyClass.SUM]} ON COLUMNS FROM [salesdata_table] CELL PROPERTIES VALUE, FORMATTED_VALUE, BACK_COLOR, FORE_COLOR, FONT_FLAGS",
       "updateMode": "once"
      },
      "serverKey": "default",
      "widgetKey": "plotly-area-chart"
     }
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "session.visualize()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizing the info in a dashboard \n",
    "\n",
    "The various scenarious and simulations can be simultaneously compared using a dashbard in atoti. \n",
    "These dashoboards are highly customisable and can be designed using the various widgets as per the business objective.\n",
    "These can be saved and shared silpky by sharing a URL.\n",
    "Click on the link below to explore a sample dashboard we created!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session.link(path=\"#/dashboard/088\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusion:\n",
    "\n",
    "Such insights from the aviation industry can help the airlines for sales and marketing, exciting offers to attract customers post COVID-19 restrictions in travel, and generate any additional insights that could benefit their business as well their customer services. \n",
    "\n",
    "For more interesting notebooks, check out the [atoti gallary](https://github.com/atoti/notebooks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# _ _The end_ _"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
